### ISSUE TITLE
`require` is weird

### ISSUE THREAD

**Issue by rybern** [swarm-game/swarm#738]
Apologies if this has already been beaten to death somewhere, or if I'm wrong, I'm a noob :)

**Is your feature request related to a problem? Please describe.**
I've found the behavior of `require` surprising in a few ways:

1. Lack referential transparency:
```
// Doesn't get a tree
let r = require 1 "tree" in build { r; h <- has "tree"; log (if h {"got a tree"} {"didn't get a tree"}) } end
// Gets a tree
build {require 1 "tree"; h <- has "tree"; log (if h {"got a tree"} {"didn't get a tree"}) } end
```
2. Can only be used with a string literal, so this causes an error:
`\s. build {require 1 s}`

3. No type signature

These lead to some issues:
- Learning curve to discover that `require` needs to be explicitly written inside `build`. Especially since `require`s compose with `cmd` actions, I expected them to have types and work like other expressions. I might have glossed over this in the tutorial.
- Inability to abstract or compose requirements. For example, it's natural to want to generalize a "planterRobot" over different kinds of items, but I can't do that very well when I have to write out the type of item wherever I call `build`

**Describe the solution you'd like**
I think `require` needs a rework. A couple possibilities:
 * Can we have `build` search through the top-level of the instructions like in my example? This would allow some abstraction/composition, although `require` expressions would still behave weirdly
 * How about a monoidal `requirement` type? Then requirements can be a second argument to `build` instead of mixing them in with the instructions

**Comment by xsebek** [comment]
Hi @rybern, thanks for reporting these issues! :+1: 

1. IIRC this is caused by `r`'s requirements being recorded only after it is used - if you used `def`, then using it later will work.
  A related issue is #540. But I may be wrong about this.
2. AFAIK that is by design. I imagine it is like a recipe in a cooking book - `require` is the list at beginning of the recipe that you have to know before going shopping or in this case `build`ing the robot.
  Maybe this is too restrictive and we could allow some variable parameters to `require` as long as the variable is known before we build the robot.
3. What do you mean? It shows `cmd unit` to me. :raised_eyebrow: 

TIP: Use `say` instead as it will show in the Messages. :wink: 

---

- Right, we could produce some meaningful warning when you use `require` outside `build`. So far we can only hope that the `require` tutorials will show you the proper usage, but we should do better.
- Agreed, I think it is simply harder to implement that way. :wink: But there is no moral reason why we should not extend it to support such usage.

--- 

- If you read through the original proposal, #201, you will see that `require` composes. I am not sure what exactly you are proposing, but I would not want to ignore `require`ments that are not at the top level. 
- That is an interesting proposal, but would break some use cases - the idea is that inner commands can specify what they need and you will not have to add it up by hand yourself. As a toy example consider this:
 `build {place1245stones; place33stonesSomewhere; placeSomeMoreStones}`

**Comment by rybern** [comment]
Thanks @xsebek!

1. This is totally the issue! I was extrapolating how `require` must work given how it works with `let`; didn't think to try `def` on a separate line
2. I'd argue for less restrictive, since requiring a literal means I need to hand-write "tree" once as a require (`require 1 "tree"`) and then again as a string. But I guess that'd be harder to implement since you'd have to check that the string is known at `build`-time. No big deal not to have this.
3. Oh, right, because it's variadic, so `require 1 "tree"` and `require "logger"` are `cmd unit` but `require` has no type

So I think it's just #540 that's tripping me up and I can close this :+1: :)

**Comment by byorgey** [comment]
I echo @xsebek 's thanks.  I agree, `require` is weird, and we need to do a better job explaining it and/or fixing some of the issues you point out.

First of all, your example re: referential transparency is most definitely a bug. `let x = a in ... x ...` and `... a ...` should always be entirely equivalent (except perhaps re: performance), as you would expect. This might be #540 as @xsebek suggests but I am not sure, it could be something else  Would you mind splitting that out as a separate issue so we can better keep track of it?

Second, I just wanted to point out that based on [the gist you linked](https://gist.github.com/rybern/4f42f178ae3b987613eabf2290d1b8d3) from #736, it looks like you are `require`ing some unnecessary/redundant things.  In many cases, writing `require` should not be necessary.  For example, you can just write  `build {move}`, you do not have to write `build {require "treads"; move}`, because capability analysis will figure out that treads are required to execute the `move` instruction.  However, there are some known bugs where writing extra `require` is necessary (e.g. #540), I don't know if you were running into that.

In any case, there are good reasons for `require` to be the way it is (i.e. having no type in and of itself, requiring a string literal, etc.).  At the very least I'm sure we need to do a better job explaining this in the tutorial, and perhaps once you understand the situation better you can help us think of some better solutions.  Let me see if I can explain it clearly.  The basic reason is that the requirements of a command are essentially *part of its type*, so we need to be able to figure out what the requirements of a command are *at typechecking time*.  This is why `require` takes only literal strings (and ints): if we allowed arbitrary expressions of type `string`, there would be no way, in general, to resolve them to concrete requirements at typechecking time.  This, in turn, is why `require` doesn't have a type: if it was a function with a type then you would be able to apply it to arbitrary expressions of the right type.  Instead, it is just special built-in syntax.
- A few thoughts on how we could make this better: perhaps we could pretend that `require` is a standalone function with a type, but also put in some special checking code to make sure only literals are passed to it.  I will have to think about how difficult/annoying that would be to do.
- Alternatively, we could make the syntax for `require` "weirder" so that it is really clear that it is some special built-in thing and not a function.  e.g. maybe something like `#require[tree]`, I dunno.

So why *can't* we resolve requirements at runtime? You mentioned having `build` scan through the given code at runtime looking for calls to `require` and dynamically resolving the requirements right then.  However, as @xsebek pointed out with an example, it is entirely by design that `require` does not have to be at the top level; it actually enables writing more modular and compositional code.  For example, we might define `def crossRiver = require "boat"; move; move; move end` and then we can use `crossRiver` as a subprocedure within other commands, and so on. If we define `def foo = crossRiver; grab; turn back; crossRiver` and then `build { ...  ; foo; ...}` that robot will automatically get a `boat` installed, because its program transitively requires a `boat`.

**Comment by rybern** [comment]
Thanks for the reply @byorgey! I snuck in a reply a minute before you. I think I'm caught up now. I was misunderstanding how restricted `require` was based on my experience with `let` and #540. The design as I now understand makes a lot more sense.

> Would you mind splitting that out as a separate issue so we can better keep track of it?

I confirmed that my example is in line with #540:
```
> let a = require 1 "tree" in build { a; h <- has "tree"; say (if h {"yep"} {"nope"}) }
"nope"
> def b = require 1 "tree" end; build { b; h <- has "tree"; say (if h {"yep"} {"nope"}) }
"nope"
> build { b; h <- has "tree"; say (if h {"yep"} {"nope"}) }
"yep"
> def c = require 1 "tree" end
> build { c; h <- has "tree"; say (if h {"yep"} {"nope"}) }
"yep"
```

> it looks like you are requireing some unnecessary/redundant things.

Yep, I added those to fix some robot crashes, and I've been constantly using `let` to build robots, so it was probably #540 again

> I'm sure we need to do a better job explaining this in the tutorial

The way `require` should work sans #540 is pretty much what I had in mind after doing the tutorials, so I think it's on point :) I've roped a friend who's not a functional programmer into trying the game and I'll report back if he finds any gaps in the tutorial.

> A few thoughts on how we could make this better..

A couple more options (that probably aren't worth it):
 * Make new types for literals, like `string_lit` and `int_lit`, that can be coerced into `string` and `int` but not the other way around, so when `require` takes a `string_lit` it's guaranteed to be knowable at `build`-time
 * Check if the arguments are knowable at `build`-time by trying to evaluate them or checking for free variables, and error out if not

Ultimately though, the string issue is no big deal now that I know `require` is intended to compose in general :)
